src/python/pants/init/logging.py

```diff
@@ -73,79 +46,120 @@ def formatException(self, exc_info):
 
-def setup_logging(global_bootstrap_options: OptionValueContainer, stderr_logging: bool) -> None:
-    """Sets up logging for a Pants run.
 
-    This is called in two contexts: 1) PantsRunner, 2) DaemonPantsRunner. In the latter case, the
-    loggers are saved and restored around this call, so in both cases it runs with no handlers
-    configured (and asserts so!).
-    """
-    if get_logging_handlers():
-        raise AssertionError("setup_logging should not be called while Handlers are installed.")
+@contextmanager
+def initialize_stdio(global_bootstrap_options: OptionValueContainer):
+    """Mutates sys.std* and logging to route stdio for a Pants process to thread local destinations.
 
-    global_level = global_bootstrap_options.level
-    log_dir = global_bootstrap_options.logdir
+    In this context, `sys.std*` and logging handlers will route through Rust code that uses
+    thread-local information to decide whether to write to a file, or to stdio file handles.
 
+    To control the stdio destination set by this method, use the `stdio_destination` context manager.
+
+    This is called in two different processes:
+    * PantsRunner, after it has determined that LocalPantsRunner will be running in process, and
+      immediately before setting a `stdio_destination` for the remainder of the run.
+    * PantsDaemon, immediately on startup. The process will then default to sending stdio to the log
+      until client connections arrive, at which point `stdio_destination` is used per-connection.
+    """
+    global_level = global_bootstrap_options.level
     log_show_rust_3rdparty = global_bootstrap_options.log_show_rust_3rdparty
     use_color = global_bootstrap_options.colors
     show_target = global_bootstrap_options.show_log_target
-    log_levels_by_target = get_log_levels_by_target(global_bootstrap_options)
+    log_levels_by_target = _get_log_levels_by_target(global_bootstrap_options)
     message_regex_filters = global_bootstrap_options.ignore_pants_warnings
-
-    init_rust_logger(
-        global_level,
-        log_show_rust_3rdparty,
-        use_color,
-        show_target,
-        log_levels_by_target,
-        message_regex_filters,
-    )
-
-    if stderr_logging:
-        setup_logging_to_stderr(global_level, global_bootstrap_options.print_stacktrace)
-
-    if log_dir:
-        setup_logging_to_file(global_level, log_dir=log_dir)
-
-
-def get_log_levels_by_target(global_bootstrap_options: OptionValueContainer) -> Dict[str, LogLevel]:
+    print_stacktrace = global_bootstrap_options.print_stacktrace
+
+    # Set the pantsd log destination.
+    # TODO: This should likely convert into a generic log, possibly with a symlink from the
+    # previous location.
+    log_path = os.path.join(global_bootstrap_options.pants_workdir, "pantsd", "pantsd.log")
+    safe_mkdir_for(log_path)
+
+    # Initialize thread-local stdio, and replace sys.std* with proxies.
+    stdin, stdout, stderr = sys.stdin, sys.stdout, sys.stderr
```

Comment:
Consider renaming to `original_stdin`. It took me a minute with line 157 to figure out what `stdin` referred to.